### PR TITLE
Add e2e fixes

### PR DESCRIPTION
**Pull request by thisac**
* Adds a small bug-fix for the "barrier" keyword, flattening the wires `anylist`.
* Adds a helper function that moves braces to separate lines so that they are equal to what the serializer outputs for the e2e tests.
* Removes whitespaces in strings before assertion (not really a good solution, but a quick one to get those tests to pass)
* Adds a test that deserializes/serializes single strings (including the specific ones that currently fail)

### CODE REVIEW

File: qdata/parser.py [hunk 1]
```diff
@@ -305,7 +305,7 @@ def qop(self, *args):
 
         if args[0] == "reset":
             # reset <argument>;
-            wires = args[1].list
+            wires = args[1]
```
Comment (thisac) [inline qdata/parser.py]:
```suggestion
            wires = args[1].list
```

File: tests/test_qasm_examples.py [hunk 2]
```diff
@@ -1,14 +1,34 @@
...
+            if len(split_lines) == 1:
+                new_line = split_lines
+            elif split_lines[1] == "":
+                new_line = [clean(split_lines[0]), b]
+            elif split_lines[0] == "":
+                new_line = [b, *fix_braces(split_lines[1:])]
+            else:
+                new_line = np.ravel([[ll, b] for ll in split_lines])[:-1]
+
+            all_lines.extend([clean(l) for l in new_line if clean(l) and clean(l) != ";"])
```
Comment (co9olguy) [inline tests/test_qasm_examples.py]: Looks like a good place ot use the walrus operator :=

File: tests/test_qasm_examples.py [hunk 3]
```diff
@@ -39,12 +59,46 @@ def test_deserialize_and_serialize(fname):
...
     tree = parser.QASMToIRTransformer().transform(tree)
     serialized_lines = tree.serialize().split("\n")
     serialized_lines = [clean(line) for line in serialized_lines]
     serialized_lines = [line for line in serialized_lines if line]
 
+    # remove all whitespaces temporarily for tests to pass; should come up with
+    # better solution
+    for i, l in enumerate(formatted_qasm_lines):
+        formatted_qasm_lines[i] = l.replace(" ", "")
+        serialized_lines[i] = serialized_lines[i].replace(" ", "")
```
Comment (co9olguy) [inline tests/test_qasm_examples.py]: This shouldn't be necessary. The whitespace is already normalized in the `clean` function. :thinking:. It might be the `fix_braces` function above? 

What happens if you repeat lines 48-49 here again?

File: tests/test_qasm_examples.py [hunk 4]
```diff
@@ -39,12 +59,46 @@ def test_deserialize_and_serialize(fname):
...
+
     assert len(formatted_qasm_lines) == len(serialized_lines)
     for (l1, l2) in zip(formatted_qasm_lines, serialized_lines):
         assert l1 == l2
+
+# tests deserializing/serializing a set of specific strings
+@pytest.mark.parametrize("stmt, expected_stmt", [
```
Comment (co9olguy) [inline tests/test_qasm_examples.py:245]: These are helpful to have! :+1:

File: tests/test_qasm_examples.py [hunk 5]
```diff
@@ -39,12 +59,46 @@ def test_deserialize_and_serialize(fname):
...
+    ("OPENQASM 2.0;\nqreg q[1];", "OPENQASM 2.0;\nqreg q[1];"),
+    ('include "qelib1.inc";', 'include "qelib1.inc";'),
+    ("gate pre q { }", "gate pre q"),
+    ("gate post q { }", "gate post q"),
+    ("gate pre q", "gate pre q"),  # fails during parsing, unless updating grammar
+    ("gate post q", "gate post q"),  # fails during parsing, unless updating grammar
```
Comment (co9olguy) [inline tests/test_qasm_examples.py]: No semicolons?

**Review by co9olguy — CHANGES_REQUESTED** [review]
Thanks @thisac for tackling this.

I'm not going to approve this version because I don't think the temporary solution to remove all whitespace is worth keeping unfortunately. We'd just have to fix it anyway

File: tests/test_qasm_examples.py (hunk 5 above)
Comment (thisac) [reply tests/test_qasm_examples.py]: Nope, the gates seem to be written as `gate abc { statement; }`, so if there's no statment, and the empty braces are removed, there's no semicolon left either. That's why I was a bit confused whether this was supported or not.

File: tests/test_qasm_examples.py (hunk 3 above)
Comment (thisac) [reply tests/test_qasm_examples.py]: Oh, I see. Unfortunately, that still doesn't solve this. The issue lies with the cases where the formatted lines differ like e.g. `if(syn==1) x q[0];`. This string would only be split at the existing whitespaces while the serialized string might look like `'if (syn == 1) x q[0];'.split()` and thus wouldn't split in the same places. Does this make sense?

File: tests/test_qasm_examples.py (hunk 2 above)
Comment (thisac) [reply tests/test_qasm_examples.py]: Oh, true. Never gotten to really use it yet. :+1:

File: tests/test_qasm_examples.py (hunk 3 above)
Comment (co9olguy) [reply tests/test_qasm_examples.py]: :thinking: hmm, wasn't able to follow

File: tests/test_qasm_examples.py (hunk 5 above)
Comment (co9olguy) [reply tests/test_qasm_examples.py]: ahh I see. In the original spec, something like `gate pre q` is simply not allowed, but `gate pre q { }` is (semicolons are allowed to be ommited if there are curly braces). Thus we will never encounter a conformant case "in the wild" which has neither a semicolon nor curly brackets.

I think the issue is not that inputs like `gate pre q` are causing failures (such a statement is not a valid part of the grammar), but it is more that removing empty brackets just needs to be handled more carefully in the test file's string cleaning functions

Comment (thisac) [reply tests/test_qasm_examples.py]: That makes more sense. Then the grammar is correct as it is. I could look into fixing this when it becomes a priority again.

File: tests/test_qasm_examples.py [hunk 6]
```diff
@@ -1,17 +1,86 @@
...
+
+def split_at_semicol(qasm_lines):
+    """Helper function to split lines at semicolons"""
+    form_lines = []
+    for line_1 in qasm_lines:
+        split_lines = line_1.split(";")
+        if len(split_lines) == 1:
+            form_lines.append(split_lines[0])
+        elif len(split_lines) == 2 and split_lines[1] == "":
+            form_lines.append(split_lines[0] + ";")
```
Comment (josh146) [inline tests/test_qasm_examples.py]: How come this elif can occur? This is the case where the line looks like `"something;"`?

Comment (thisac) [reply tests/test_qasm_examples.py]: Yes, it should happen every time there's a correct statement that ends with a `;` (e.g. `something;`). But looking at it now it seems like it should be covered by the following `else`. :thinking:

File: tests/test_qasm_examples.py [hunk 7]
```diff
@@ -1,17 +1,86 @@
...
+def fix_braces(qasm_lines):
+    """Helper function to place braces on separate lines"""
+    form_lines = []
+    for line_1 in qasm_lines:
+        res = line_1.split("{")
+        next_qasm_lines = [clean(em) for el in list(zip(["{"]*len(res), res)) for em in el if clean(em)][1:]
+        for line_2 in next_qasm_lines:
+            res = line_2.split("}")
+            lines = [clean(em) for el in list(zip(["}"]*len(res), res)) for em in el if clean(em)][1:]
+            form_lines.extend(lines)
```
Comment (josh146) [inline tests/test_qasm_examples.py]: This is also quite difficult to parse 😆

File: qdata/ir.py [hunk 8]
```diff
@@ -389,6 +389,9 @@ def __init__(self, op):
         super().__init__(decl_type=Declarations.GATE, op=op)
 
     def declaration_str(self):
+        if not self.goplist:
+            return f"gate {self.kwargs['op']}"[:-1] + " { }"
```
Comment (josh146) [inline qdata/ir.py:393]: What was the bug/edge case that required this change?

File: qdata/parser.py [hunk 9]
```diff
@@ -302,14 +302,15 @@ def qop(self, *args):
         if len(args) == 1:
             # <uop>
             return args[0]
-
         if args[0] == "reset":
             # reset <argument>;
-            wires = args[1]
+            # `Op` takes a list of wires while `args[1]`` is of type `['q', 0]``
```
Comment (josh146) [inline qdata/parser.py]:
```suggestion
            # `Op` takes a list of wires while `args[1]` is of type `['q', 0]`
```

File: tests/test_qasm_examples.py (hunk 6 above)
Comment (josh146) [reply tests/test_qasm_examples.py]: Yes, that appears to be true

File: tests/test_qasm_examples.py [hunk 10]
```diff
@@ -1,17 +1,136 @@
...
+        if len(split_lines) == 1:
+            form_lines.append(split_lines[0])
+        elif len(split_lines) == 2 and split_lines[1] == "":
+            form_lines.append(split_lines[0] + ";")
+        else:
+            for line_2 in split_lines:
+                delim = ";" * (clean(line_2) not in ["{", "}"])
+                if clean(line_2):
+                    form_lines.append(clean(line_2) + delim)
+    return form_lines
```
Comment (josh146) [inline tests/test_qasm_examples.py:36]: I think this helper function can be improved by commenting each if branch, to explain what we are branching at, and the required fix

File: tests/test_qasm_examples.py [hunk 11]
```diff
@@ -1,17 +1,136 @@
...
+
+def fix_braces(qasm_lines):
+    """Helper function to place braces on separate lines"""
+    form_lines = []
+    for line_1 in qasm_lines:
+        # if a line contains empty braces, format them separately
+        match = re.search(r"{\s*}", line_1)
+        if match:
+            form_lines.append(line_1.replace(match.group(), "{ }"))
+            continue
```
Comment (josh146) [inline tests/test_qasm_examples.py]: Just to double check I am following this properly: this part does `{    <arbitrary whitespace>  } --> { }`?

File: tests/test_qasm_examples.py [hunk 12]
```diff
@@ -1,17 +1,136 @@
...
+
+        # split at "{" and re-add them as separate list-elements
+        res = line_1.split("{")
+        next_qasm_lines = [clean(em) for el in list(zip(["{"]*len(res), res)) for em in el if clean(em)][1:]
+        for line_2 in next_qasm_lines:
+            # split the new list at "}" and re-add them as separate list-elements
+            res = line_2.split("}")
+            lines = [clean(em) for el in list(zip(["}"]*len(res), res)) for em in el if clean(em)][1:]
+            form_lines.extend(lines)
```
Comment (josh146) [inline tests/test_qasm_examples.py]: For this section, is the overall logic

Comment (josh146) [reply tests/test_qasm_examples.py]: Actually, I'm a bit lost at the list comprehension. I find it a bit difficult to understand the comments as well

File: tests/test_qasm_examples.py [hunk 13]
```diff
@@ -1,17 +1,136 @@
...
+
+    #asdad
```
Comment (josh146) [inline tests/test_qasm_examples.py]: Comment TBD? 😆

File: tests/test_qasm_examples.py [hunk 14]
```diff
@@ -1,17 +1,136 @@
...
+    tab = False
+    for i in range(1, len(form_lines) - 1):
+        if tab and form_lines[i] != "}":
+            form_lines[i] = "    " + form_lines[i]
+
+        if form_lines[i] == "{":
+            tab = True
+        elif form_lines[i] == "}":
+            tab = False
```
Comment (josh146) [inline tests/test_qasm_examples.py]: This variable could be named a bit better, perhaps `indented_block`. That way, the if statement would be `if indented_block and form_lines...`

File: tests/test_qasm_examples.py [hunk 15]
```diff
@@ -1,17 +1,136 @@
...
+
+def fix_ifs(line):
+    """Helper function to format if-statements correctly"""
+    p = re.compile(r"(if)\s*\(\s*(\w+)\s*(==|!=|>=|<=|<|>)\s*(\w+)\s*\)")
+    match = p.match(line)
+    # if an if-string is found, "build up" a correctly formatted string
+    if match:
+        ifstring = match.group(1) + " (" + match.group(2) + " " + match.group(3) + " " + match.group(4) + ")"
+        return ifstring + line[match.span()[1]:]
+    return line
```
Comment (josh146) [inline tests/test_qasm_examples.py:119]: :+1: This one I could follow 😆

Comment (josh146) [reply tests/test_qasm_examples.py:119]: Although I would recommend adding a more detailed docstring to make it easy to understand _without_ looking at the code what is happening. E.g.,

```
"""Normalize spacing in if statements.
The normalized if statement will have the form:

.. code-block:: c

    if (name condition name)

with the whitespace exactly as indicated.
"""
```

The comment could also be expanded (to describe the capture groups, etc.). 'correctly formatted string' is not very descriptive, since a 'correct' if statement string is never defined.

File: tests/test_qasm_examples.py [hunk 16]
```diff
@@ -1,17 +1,136 @@
...
+
+
+def fix_spaces(line):
+    """Helper function to format spaces before/after paranthesis and after commas correctly"""
+    line = ",".join(re.split(r"\s*,\s*", line))
+    line = "(".join(re.split(r"\s*\(\s*", line))
+    line = ") ".join(re.split(r"\s*\)\s*", line))
+    return clean(line)
```
Comment (josh146) [inline tests/test_qasm_examples.py]: Was this @co9olguy original space normalizing logic? Or is this new?

File: tests/test_qasm_examples.py [hunk 17]
```diff
@@ -1,17 +1,136 @@
...
+
+
+# for variable names that are used by Python, and cannot be used as SymPy symbols
+invalid_var_names = ["lambda"]
+def fix_math_expr(line):
+    """Helper function to format math expressions correctly"""
```
Comment (josh146) [inline tests/test_qasm_examples.py]: Docstring could be more detailed here (especially defining what 'correct' means in this case! Or even detailing why this function is needed)

File: tests/test_qasm_examples.py [hunk 18]
```diff
@@ -1,17 +1,136 @@
...
+
+    # remove the if-parts from the string (doens't support maths in if-statments)
+    if "if" in line:
+        line = re.split(r"if\s*\(.+?\)\s*", line)[-1]
```
Comment (josh146) [inline tests/test_qasm_examples.py]: Is this needed to extract potential maths from the if condition?

File: tests/test_qasm_examples.py [hunk 19]
```diff
@@ -1,17 +1,136 @@
...
+
+    p = re.compile(r"\(.+\)")
```
Comment (josh146) [inline tests/test_qasm_examples.py:157]: This regex looks like it just matches anything?

File: tests/test_qasm_examples.py [hunk 20]
```diff
@@ -1,17 +1,136 @@
...
+        p_match = res.group()
+        free_params = re.findall(r"[a-zA-Z]\w*", p_match)
+
+        for i, p in enumerate(free_params):
+            if p in invalid_var_names:
+                free_params[i] = p + "_var"
+                p_match = p_match.replace(p, free_params[i])
+                invalid_var_names_list.append((p, free_params[i]))
+
+        sympy.var(free_params)
```
Comment (josh146) [inline tests/test_qasm_examples.py]: Does this line do anything?

File: tests/test_qasm_examples.py [hunk 21]
```diff
@@ -1,17 +1,136 @@
...
+        p_replace = p_match
+        if p_replace[0] == "(":
+            p_replace = p_replace[1:-1]
+        p_replace = ",".join([str(eval(r)) for r in p_replace.split(",")])
+        for var in invalid_var_names_list:
+            p_replace = p_replace.replace(var[1], var[0])
+            p_match = p_match.replace(var[1], var[0])
+        line = line.replace(p_match, "(" + p_replace + ")")
+
+    return line
```
Comment (josh146) [inline tests/test_qasm_examples.py]: This is _definitely_ worth commenting, I got very lost! 😆

**Review by josh146 — COMMENTED** [review]
Hey @thisac, great job getting all the tests to pass! This is the first time getting the green tick on the e2e tests which is amazing.

However, I had a bit of trouble following (a) what the formatting functions are doing, and (b) why some of the formatter functions were needed. Some (such as the math formatting) you explained offline to me, but it would be great to include those explanations in the functions themselves.

I can imagine a case where we change the syntax in the future, and need to update/change these formatting functions - and have forgotten how they work or what they do!

(Sidenote: I still suggest we simply avoid the complexity of code formatting by simply running `clang-format` on the input file and the output file)

File: tests/test_qasm_examples.py (hunk 12 above)
Comment (josh146) [reply tests/test_qasm_examples.py]: re.split("([^{\s+]+|{)", "a { bla bla")

File: qdata/ir.py (hunk 8 above)
Comment (co9olguy) [reply qdata/ir.py:393]: Looks like it will just force a declaration to always have empty brackets in the case where there is no actual operations to define inside?

File: tests/test_qasm_examples.py [hunk 22]
```diff
@@ -1,17 +1,136 @@
 import os
 import re
+import sympy
```
Comment (co9olguy) [inline tests/test_qasm_examples.py:3]: :crying_cat_face:

File: tests/test_qasm_examples.py (hunk 16 above)
Comment (co9olguy) [reply tests/test_qasm_examples.py]: My functions would just turn any whitespace into a single space. I think this is to insert whitespace in places where it might not be explicitly?

File: tests/test_qasm_examples.py [hunk 23]
```diff
@@ -1,17 +1,136 @@
...
+def format_code(qasm_lines):
+    """Format the code"""
+    formatted_qasm_lines = [clean(line) for line in qasm_lines if clean(line)]
+    formatted_qasm_lines = split_at_semicol(formatted_qasm_lines)
+    formatted_qasm_lines = [fix_spaces(line) for line in formatted_qasm_lines]
+    formatted_qasm_lines = [fix_ifs(line) for line in formatted_qasm_lines]
+    formatted_qasm_lines = fix_braces(formatted_qasm_lines)
+    formatted_qasm_lines = [fix_math_expr(line) for line in formatted_qasm_lines]
+
+    return formatted_qasm_lines
```
Comment (co9olguy) [inline tests/test_qasm_examples.py:218]: Despite how scary some of the above helper functions look, it's nice to see things here so clearly.

File: tests/test_qasm_examples.py [hunk 24]
```diff
@@ -24,27 +143,36 @@ def test_deserialize_and_serialize(fname):
...
+    ("measure q[0] -> c[0];", "measure q[0] -> c[0];"),
+    ("reset q[0];", "reset q[0];"),
+    ("if(c==3) u1(pi/2+pi/4) q[2];", "if (c == 3) u1(3*pi/4) q[2];"),  # note the different args and whitespace
+])
+def test_lines(stmt, expected_stmt):
+    """TODO"""
+    tree = parser.qasm_parser.parse(stmt)
+    tree = parser.QASMToIRTransformer().transform(tree)
+    serialized_stmt = tree.serialize()
+    assert serialized_stmt == expected_stmt
```
Comment (co9olguy) [inline tests/test_qasm_examples.py:265]: I have to admit, I'm surprised at how many new "cleanup" functions needed to be added to make things work. I had really figured it would not require such deep effort. Maybe I should have been more accepting of the suggestion to just remove all spaces :sweat:

File: tests/test_qasm_examples.py (hunk 16 above)
Comment (thisac) [reply tests/test_qasm_examples.py]: It inserts/removes whitespaces in front of and behind commas and parantheses so that it always looks like e.g. `u(a,b,c)` and _never_ `u (a,b, c )`. We might want to change these rules, although what they do now is making sure it's formatted the same way as they're currently output.

File: tests/test_qasm_examples.py (hunk 18 above)
Comment (thisac) [reply tests/test_qasm_examples.py]: The other way around; this simply ignores any parantheses connected to if-statements.

File: tests/test_qasm_examples.py (hunk 19 above)
Comment (thisac) [reply tests/test_qasm_examples.py:157]: It does! It matches a parantheses pair. I'm assuming here that there will always only be one pair per line (except the if-statement, which we ignore earlier), although it works with nested parantheses.

File: tests/test_qasm_examples.py (hunk 20 above)
Comment (thisac) [reply tests/test_qasm_examples.py]: Yes, it makes the names in `free_params` symbolic.

File: qdata/ir.py (hunk 8 above)
Comment (josh146) [reply qdata/ir.py:393]: Oh, that makes sense

File: tests/test_qasm_examples.py (hunk 19 above)
Comment (josh146) [reply tests/test_qasm_examples.py:157]: My mistake, I read it as `(.+)` instead of `\(.+\)` 😆 

`sed` has trained me to ignore leading/training slashes.

File: tests/test_qasm_examples.py (hunk 24 above)
Comment (josh146) [reply tests/test_qasm_examples.py:265]: @co9olguy I'm very much in favour of returning to remove all spaces for the purposes of the tests (or, using an external C-like formatter). @thisac has done a great job hunting down all the formatting edge cases in these formatting functions! But I feel that the complexity of these formatting normalization functions will be 

* very hard to maintain,

* likely still contain small bugs/edge cases simply due to the limited example data,

* result in non-trivial developer overhead when writing new tests or changing/tweaking the grammar/syntax.

Simply removing whitespace will allow us to develop faster, and provide more confidence in the test suite.

File: tests/test_qasm_examples.py [hunk 25]
```diff
@@ -8,57 +8,108 @@
...
 
 def split_at_semicol(qasm_lines):
-    """Helper function to split lines at semicolons"""
+    """Helper function to split lines at semicolons.
+
+    * If there are no ``;`` in the line, it re-appends it
+    * If there's a single ``;`` at the end of the line, it appends the
+      split line and adds back the ``;``.
+    * If there's at least one ``;`` in the middle of the line, it appends
+      each split section if the line is neither ``"{"`` nor ``"}"``.
```
Comment (josh146) [inline tests/test_qasm_examples.py:17]: This is really useful/helpful!

File: tests/test_qasm_examples.py [hunk 26]
```diff
@@ -8,57 +8,108 @@
...
+            # a single ``;`` at the end of the line, then append the line + ";"
             form_lines.append(split_lines[0] + ";")
         else:
+            # at least one ``;``, then append each sub-string + ";" if it's neither
+            # ``"{"`` nor ``"}"``, else only append sub-string.
             for line_2 in split_lines:
-                delim = ";" * (clean(line_2) not in ["{", "}"])
-                if clean(line_2):
-                    form_lines.append(clean(line_2) + delim)
+                # delim should be "" if line is not "{" or "}"
```
Comment (josh146) [inline tests/test_qasm_examples.py:32]: 🎉

File: tests/test_qasm_examples.py [hunk 27]
```diff
@@ -8,57 +8,108 @@
...
+        start = m.start()
+        # add the preceding part and delim as two different list elements
+        split_list.extend([line[last_end:start].strip(), delim])
+        last_end = m.end()
+
+    # add the final part of the string as the final list element
+    if line[last_end:].strip():
+        split_list.append(line[last_end:].strip())
+
+    return split_list
```
Comment (josh146) [inline tests/test_qasm_examples.py]: This is so much more readable 💯 🏆

File: tests/test_qasm_examples.py [hunk 28]
```diff
@@ -8,57 +8,108 @@
...
+
+
 def fix_braces(qasm_lines):
-    """Helper function to place braces on separate lines"""
+    """Helper function to place braces on separate lines.
+
+    Assuming there are only closed braces (e.g. no single braces without a
```
Comment (josh146) [inline tests/test_qasm_examples.py]: closing -> matching (for some reason, GitHub won't let me make a suggestion)

File: tests/test_qasm_examples.py [hunk 29]
```diff
@@ -8,57 +8,108 @@
...
+    partner), each brace will be split up and added as a separate element
+    in the returned list. For example, ``["gate { a; }"]`` will become
+    ``["gate", "{", "a;", "}"]``.
```
Comment (josh146) [inline tests/test_qasm_examples.py:74]: 🥇

File: tests/test_qasm_examples.py [hunk 30]
```diff
@@ -69,58 +120,94 @@ def fix_ifs(line):
...
-    # remove the if-parts from the string (doens't support maths in if-statments)
+    Checks for closed parantheses in a line from a qasm-file (ignoring
+    if-statements) and re-formats the mathematical expression(s) inside of it
+    using SymPy. For example, ``((alpha + beta) / 2)`` becomes ``(alpha/2 +
+    beta/2)``, and ``(4*pi/(1+3))`` becomes ``(pi)``.
+
+    Assumptions/caveats:
+
+    * Only one pair of parantheses are assumed per line (excluding if-statments).
+    * All parantheses are assumed to contain mathematical expressions.
```
Comment (josh146) [inline tests/test_qasm_examples.py:150]: This function I think is the one that worries me the most, in that changes in the grammar/example data _could_ require maintenance here.

File: tests/test_qasm_examples.py [hunk 31]
```diff
@@ -69,58 +120,94 @@ def fix_ifs(line):
...
+        if p in invalid_var_names:
+            free_params[i] = p + "_var"
+            p_match = p_match.replace(p, free_params[i])
+            invalid_var_names_list.append((p, free_params[i]))
+
+    # create symbols and inject them into the global namespace
+    sympy.var(free_params)
+
+    # extract and evaluate the mathematical expressions inside of the
+    # parantheses; e.g., (4*a/2, b, 1+2) becomes (2*a, b, 3).
```
Comment (josh146) [inline tests/test_qasm_examples.py:178]: These comments are great 🙌

File: tests/test_qasm_examples.py [hunk 32]
```diff
@@ -69,58 +120,94 @@ def fix_ifs(line):
...
 invalid_var_names = ["lambda"]
 def fix_math_expr(line):
-    """Helper function to format math expressions correctly"""
+    """Helper function to format math expressions
 
-    # remove the if-parts from the string (doens't support maths in if-statments)
+    Checks for closed parantheses in a line from a qasm-file (ignoring
+    if-statements) and re-formats the mathematical expression(s) inside of it
+    using SymPy. For example, ``((alpha + beta) / 2)`` becomes ``(alpha/2 +
+    beta/2)``, and ``(4*pi/(1+3))`` becomes ``(pi)``.
```
Comment (josh146) [inline tests/test_qasm_examples.py:145]: Are there never cases where math occurs not within matching parenthesis?

File: tests/test_qasm_examples.py (hunk 28 above)
Comment (josh146) [reply tests/test_qasm_examples.py]:
```suggestion
    Assuming there are only matching braces (e.g. no single braces without a
```

**Review by josh146 — APPROVED** [review]
Thanks for going back and cleaning up the formatting functions @thisac! I find them easier to follow and understand now.

Even with the complexity of the formatting functions, merging this in will be a big net positive as it allows the test suite to now pass

File: tests/test_qasm_examples.py (hunk 30 above)
Comment (thisac) [reply tests/test_qasm_examples.py:150]: I agree. This is probably the most complex thing that I'm doing here. :thinking:

File: qdata/parser.py (hunk 9 above)
Comment (co9olguy) [reply qdata/parser.py]: @thisac don't forget to merge this suggestion

File: tests/test_qasm_examples.py (hunk 32 above)
Comment (thisac) [reply tests/test_qasm_examples.py:145]: I'm not absolutely sure, but it seems like there aren't. That's an assumption I've made here to make things a bit easier (although it's not fully bulletproof, and it might not be the best way of doing things if we want to expand what's allowed in the future either).

File: tests/test_qasm_examples.py [hunk 33]
```diff
@@ -1,17 +1,223 @@
...
+
+    Args:
+        line (str): the string to split
+        delim (str): the delimiter to split at
+
+    Returns:
+        list: all substrings and delimiters (as separate elements)
+    """
+    if line == delim:
+        return line
```
Comment (co9olguy) [inline tests/test_qasm_examples.py]: This conditional seems to lead to a different return type than listed in the docstring

**Review by co9olguy — APPROVED** [review]

File: tests/test_qasm_examples.py [hunk 34]
```diff
@@ -24,27 +230,36 @@ def test_deserialize_and_serialize(fname):
...
+    ("pre q[0];", "pre q[0];"),
+    ("barrier q;", "barrier q;"),
+    ("h q[0];", "h q[0];"),
+    ("post q[0];", "post q[0];"),
+    ("measure q[0] -> c[0];", "measure q[0] -> c[0];"),
+    ("reset q[0];", "reset q[0];"),
+    ("if(c==3) u1(pi/2+pi/4) q[2];", "if (c == 3) u1(3*pi/4) q[2];"),  # note the different args and whitespace
+])
+def test_lines(stmt, expected_stmt):
+    """TODO"""
```
Comment (co9olguy) [inline tests/test_qasm_examples.py]: docstring needs updating?

**Comment by co9olguy** [comment]
I agree with Josh that it will be nice to have the tests pass :100:. Thanks for making it all work out @thisac.
Thinking a bit bigger, having now seen your heroic effort and the complexity of all the formatters needed in the tests, if we want to continue to evolve the spec in the future, it might make more sense to either use a standard formatter or completely remove whitespace (which I originally thought of as a hack, but have now seen the light! :angel:), as have already been suggested

File: tests/test_qasm_examples.py (hunk 33 above)
Comment (thisac) [reply tests/test_qasm_examples.py]: Oh, good catch! I'll fix it!

File: qdata/parser.py (hunk 9 above)
Comment (thisac) [reply qdata/parser.py]: Oh, right, thanks for reminding me. :smile:
